Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py

```diff
@@ -0,0 +1,155 @@
+""" polls data from DS to azure logs """
+import logging
+from . import DS_api
+from . import AS_api
+from .state_serializer import State
+import json
+from . import constant
+
+logger = logging.getLogger("DS_poller")
+
+class poller:
+
+    def __init__(self, ds_id, ds_key, secret, as_id, as_key, connection_string, historical_days, url):
+        """ 
+            initializes all necessary variables from other classes for polling 
+        """
+        
+        self.DS_obj = DS_api.api(ds_id, ds_key, secret, url)
+        self.AS_obj = AS_api.logs_api(as_id, as_key)
+        self.date = State(connection_string)
+        logger.info("got inside the poller code")
+        self.event = self.date.get_last_event(historical_days)
+        if(isinstance(self.event, tuple)):
+            self.after_time = self.event[0]
+            self.before_time = self.event[1]
+            logger.info("From time: %s", self.after_time)
+            logger.info("to time: %s", self.before_time)
+        else:
+            logger.info("Polling from event number " + str(self.event))
+
+    def parse_desc(self, data):
+        """
+            adds more newlines to description for good display on Azure sentinel incidents
+        """
+        arr = data.splitlines()
+        res = ""
+        for e in arr:
+            res = res + e + "\n\n"
+        return res
+
+
+    def post_azure(self, response, item):
+        """
+            posts to azure after appending triage information on it
+        """
+        json_obj = response.json()
+        for it in item:
+            for i in range(len(json_obj)):
+                if(it['source']['alert-id'] == json_obj[i]['id'] or it['source']['incident-id'] == json_obj[i]['id']):
+
+                    comment_data = self.DS_obj.get_triage_comments(it['id'])
+                    json_obj[i]['status'] = it['state']
+                    json_obj[i]['triage_id'] = it['id']
+                    json_obj[i]['triage_raised_time'] = it['raised']
+                    json_obj[i]['triage_updated_time'] = it['updated']
```

Comment:
i might've done this flow differently - rather than adding data onto the `alert_or_incidents` object I would've created a new object with data copied onto it:

```py
azure_obj = {
  **alert_or_incident,
  'status': triage_item['state'],
  'triage_id': triage_item['id'] # etc for others
}
```

This avoids mutating the object passed into the function, which could have side-effects if the caller tries to use that object for something else later.